benchmark/README.md

```diff
@@ -0,0 +1,199 @@
+2. the secondary condition, a hard limit of submission rate, just gives user a little bit more control of the search.
+
+```
+// bm: Benchmarker
+// generator: impl LoadGenerator
+let rate = lower_bound;
+while rate < upper_bound {
+    // Generate and send transactions @ constant rate
+    let (num_submitted, num_committed, throughput) =
+        run_benchmarker_at_const_rate(bm, generator, accounts, rate);
+
+    // Measure throughput & commit ratio over several epochs
+    max_throughput = max(max_throughput, throughput);
+    if (num_committed / num_submitted) < COMMIT_RATIO_THRESHOLD {
+        break;
+    }
+    rate += inc_step
+}
+```
+
+The result we used for testing Libra at a fixed submission rate is the aggregated results over several epochs,
+and the maximum throughput will be reported when search terminates.
+
+If we dive in a little bit to the testing procedure, within each epoch, we will first generate new transactions from account chunks so that any account doesn't create more than the `capacity_per_user` limit. After measuring the throughput for several epochs, we summarize the results as total number of submitted transactions, committed transactions and overall throughput, that is the total number of committed transactions divided by total submitting and waiting durations.
+
+```
+fn run_benchmarker_at_const_rate() {
+    for _ in 0..num_epochs {
+        // Generate new accounts
+        let new_accounts = generator.gen_accounts();
+        accounts.extend(new_accounts);
+        // Generate new transactions from chunked accounts
+        for account_chunk in accounts.chunks(batch_size) {
+            let req_chunk = generator.generate_request(account_chunk);
+            requests.extend(req_chunk.into_iter());
+        }
+        result.push(bm.measure_throughput(requests, accounts, rate));
+    }
+    // Return #total submitted transactions, #total committed transactions, overall throughput.
+    aggregated_result(result)
+}
+```
+
+A side benefit of linear search is that we can observe how the entire system react to constantly injected workload. But you can also argue it may impose too much load during the search. In addition, searching usually takes very long time.
+
+
+## Denial of Service Attack on Libra
+
+To advertise the benchmarking library, this section describe how we DoS the Libra blockchain network running on bench environment. First, we need to admit that launching denial-of-service attack to Libra with transaction requests is certainly not a trivial task. Even if you can generate valid and signed transactions on your own very efficiently, AC will not let these transactions pass through because it limits the number of transactions on a per account basis. In addition, mempool also has sufficient capacity to hold a large number of not-ready transactions and even garbage collect expired transactions. Consensus also processes mempool's transactions at its own speed.
```

Comment:
> I'd remove "bench environment", because it applies to all local and some remote runs.
> 

Ok. I was not clear but mean to refer bench environment for the assumption that there is no per ip filter.

> > AC will not let these transactions pass through because it limits the number of transactions on a per account basis
> 
> Mempool, for pending transactions.

Yes, mempool is more accurate.

> But this is an arbitrary limit, and arbitrary limitation (i.e., create more accounts). Mempool's capacity can be a different type of attack.

Mentioned this point in the doc.

> 
> The core difference is that submitting a transaction (as long as we do not have batch API) has an upper cost. The cost of big `UpdateToLatestLedger` has a higher upper cost.

Mentioned this point in the doc.